### PR TITLE
Add Block Generator as bgenerator

### PR DESCRIPTION
**Pull request by myzinsky**
Added a Block Generator as used in German HAM-Radio Exam: 

image

Its number 3 in the pic.

### CODE REVIEW

**Comment by myzinsky** [comment]
@Rmano please check if I did that correctly :-)

**Comment by Rmano** [comment]
I'd check it asap; at first glance seems ok. Thanks.

File: doc/circuitikzmanual.tex [hunk 1]
```diff
@@ -3464,6 +3464,7 @@ \subsection{Block diagram components}
     \circuitdescbip*{highpass}{highpass}{}
     \circuitdescbip*{lowpass}{lowpass}{}
     \circuitdescbip*{allpass}{allpass}{}
+    \circuitdescbip*{bgenerator}{bgenerator}{}
```
Comment (Rmano) [inline doc/circuitikzmanual.tex:3467]: Please use, for example: 
`\circuitdescbip*{bgenerator}{Block generator (use \texttt{t=\dots} to specify text)}{}`

**Comment by Rmano** [comment]
Apart from the single (minor) comment, it's perfect.  I'll merge as soon as you want. 

Let me share with you my boilerplate code to check the blocks; it can be handy because that way, you can also check the code for the manual without having to recompile it every time. Clearly if you have the blocks in your local copy of `circuitikikz` you do not need to put them into the `\makeatletter`...`\makeatother` block here.

```latex
\documentclass[]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage[siunitx, RPvoltages]{circuitikz}
\usepackage{ctikzmanutils}

\makeatletter
\ctikzset{bipoles/bgenerator/width/.initial=.7}
\def\pgfcirc@twoport@draw@narrowsine#1{% #1 -> y shift
    \pgfscope
        \pgfsetlinewidth{\pgfstartlinewidth}
        \pgftransformyshift{#1\pgf@circ@res@step}
        \pgfpathmoveto{\pgfpoint{.125\pgf@circ@res@step}{0\pgf@circ@res@step}}
        \pgfpathsine{\pgfpoint{.125\pgf@circ@res@step}{.125\pgf@circ@res@step}}
        \pgfpathcosine{\pgfpoint{.125\pgf@circ@res@step}{-.125\pgf@circ@res@step}}
        \pgfpathsine{\pgfpoint{.125\pgf@circ@res@step}{-.125\pgf@circ@res@step}}
        \pgfpathcosine{\pgfpoint{.125\pgf@circ@res@step}{.125\pgf@circ@res@step}}
        \pgfusepath{draw}
    \endpgfscope
}
%% Block Generator
\pgfcirc@define@twoports{blocks}
{}
{\ctikzvalof{bipoles/bgenerator/width}}
{bgenerator}
{\ctikzvalof{bipoles/bgenerator/width}}
{\ctikzvalof{bipoles/bgenerator/width}}
{
    \pgfcirc@twoport@draw@narrowsine{0.25}{0}
    \pgfcirc@twoport@draw@narrowsine{0.0}{0}
    \pgfcirc@twoport@draw@narrowsine{-0.25}{0}
    \pgftext[x=0.45\pgf@circ@res@left]{\ctikzvalof{bipoles/twoport/text}}
}
\pgfcirc@activate@bipole@simple{l}{bgenerator}
\makeatother

\begin{document}
\section{Dummy sec}
\subsection{Dummy subsec}
\subsubsection{Dummy subsubsec}

\begin{groupdesc}
     \circuitdescbip*{bgenerator}{Block generator (use \texttt{t=\dots} to specify text)}{}
\end{groupdesc}

Example:

\begin{LTXexample}[varwidth=true]
\begin{tikzpicture}[scale=0.9]
    \ctikzset{blocks dc segments=3}
    \draw (0,2) to[sacdc] ++(2,0)
        to[bgenerator, t=G] ++(2,0);
\end{tikzpicture}
\end{LTXexample}

Check blocks orientation:

\foreach \bl in {sacdc, bgenerator} {
    \begin{tikzpicture}[]
        \draw (0,0) to[\bl, t=P, >] ++(3,0)
            to[\bl, t=P, >] ++(0,-3)
            to[\bl, t=P, >] ++(-3,0)
            to[\bl, t=P, >] ++(0,3);
    \end{tikzpicture}\qquad
}

\end{document}
```

image

**Comment by Rmano** [comment]
BTW. if you want to add your other blocks to this PR, feel free to do it. Whichever way you like. And, add something to the CHANGELOG.md file and some credits in the main documentation (like a footnote like I did for the tail-less mic, maybe?)

**Comment by Rmano** [comment]
> I will add the other blocks step by step then!

Yep, perfect. Tell me when it's ready for review/merging in.

**Comment by myzinsky** [comment]
@Rmano I have a question: 

The Crystal Generator that I added makes some problems if I scale the picture i.e. in a column:

image

The bars of the crystal symbol are removed... any ideas?

**Comment by Rmano** [comment]
Can you please post the code? I had not (yet) looked at the symbols...

**Comment by Rmano** [comment]
> The code is in [bc404b5](https://github.com/circuitikz/circuitikz/pull/693/commits/bc404b5dc40d5785488608bbc097a8726d1bf7e3)

No, I meant the code that triggers the error, the user-level one.

**Comment by myzinsky** [comment]
Unfortunately I cannot share a minimal example. When I use it out of context in a standalone it looks finde. Only in my environment that generates the latex code it does not work :-(

**Comment by Rmano** [comment]
Ah: you have `\pgfpathlineto{...}` without a `\pgfpathmoveto{...}`. I think that after a `\pgfpathuse{}` you have to set the current point. Check if that solves the problem.

**Comment by Rmano** [comment]
Hi @myzinsky  --- I was planning for a release this weekend, lot of things have accumulated. Should I wait for this PR or do you think it's better to leave it for 1.6.2?

**Comment by myzinsky** [comment]
I think I will add some more block diagrams. So lets keep it for the next release then!

**Comment by Rmano** [comment]
Ok. Did you check the `\pgfpathmoveto{...}` comment? That can create glitches that depend on the PDF rendered...

**Comment by myzinsky** [comment]
Yes sorry ... just copy paste problem :-)

**Comment by Rmano** [comment]
Hi @myzinsky --- I will do a new release soon. Any news from here?

**Comment by myzinsky** [comment]
I have now many components for our project I have to PR-them here. I think best would be if you do this release and I will push them for the next one!

**Comment by Rmano** [comment]
Fine with me!

**Comment by Rmano** [comment]
@myzinsky  Ping? 😉

**Comment by myzinsky** [comment]
Oh sorry, we got lost with our project a bit :-D But now I have time to contribute the rest :-)

**Comment by myzinsky** [comment]
These are the components that I want to contribute back: 

image

how should we proceed? Shall I just add the components to the files where I think they belong to?

**Comment by myzinsky** [comment]
This would be an example for the docs, but I dont know where to put it, maybe you could decide this:

```
% Author: Prof. Dr. Matthias Jung
% Year: 2023
\begin{tikzpicture}
    %\useasboundingbox (-3.75,-1.0) rectangle (10.5,3);
    \node[draw, powershape] at (0,0) (supply) {};
    \draw(supply.south) node[anchor=north]{\qty{13,8}{V}};
    \node[draw, trxshape] at (5,0) (trx) {};
    \draw(trx.east)
        -- ++(1,0)
        node[swrshape, anchor=west](swr){};
    \draw(swr.east)
        -- ++(1,0)
        node[antenna]{};
    \draw(trx.south) node[ground]{};
    %
    \draw[red](supply.plus)
        to [fuse, color=red] ++(2.5,0)
        to [short, i={I}, color=red] (trx.plus);
    \draw(trx.minus) to [short, i={}] (supply.minus);
    \draw(supply.u1) to [short, -o] ++(-1,0) coordinate(c1);
    \draw(supply.u2) to [short, -o] ++(-1,0) coordinate(c2);
    \draw(c1) to [open, l_={$\sim\qty{230}{\volt}$}] (c2);
\end{tikzpicture}
```

image

**Comment by myzinsky** [comment]
Another example which could be used: 

```
% Author: Prof. Dr. Matthias Jung, DL9MJ
% Year: 2023
\begin{circuitikz}[european]
    \draw(0,0) node [dinantenna](ant1){};
    \draw[thick](ant1.south) -- ++(0,-1) node [anchor=south west](){$f=\qty{145,9}{\mega\hertz}$};
    %
    \draw(3.5,0) node [dinantenna](ant2){};
    \draw(3.75,-0.5) node[tvshape, anchor=west, fill=DARCgray](tv){};
    \draw[thick] (ant2) |- (tv);
    \draw(tv.north) node[above](){TV};
    %
    \draw[DARCgray](ant1.right) node[harmonics, xscale=2*\getDarcImageFactor, yscale=1*\getDarcImageFactor, anchor=left](){};
    \draw[thick](ant1.right) node[waves, xscale=2*\getDarcImageFactor, yscale=1*\getDarcImageFactor, anchor=left](){};
    %
    \draw(tv.south) ++(0,-0.25) node[below](foo){$f=$\,\qtyrange{582}{590}{\mega\hertz}};
    \draw(foo.south) node[below](){DVB-T2 Kanal 35};
\end{circuitikz}
```

image

**Comment by myzinsky** [comment]
I wasn't quite sure how to document the solar cell. Maybe you could handle that? I hope you like the components and that many others will find them useful.

**Comment by Rmano** [comment]
They are nice! Just a couple of comments:

1. I think the cross on the SWR meter is not centered --- is it on purpose?
2. the solar cell should use the opto-arrows like photodiodes, no? 

I am a bit busy vacationing ;-) with the family now, but I would like to do a release before January 6th, so as soon as I have time I'll merge it. About the examples, add them where you think better  --- usually, when explaining the anchors, it is usually the best place.

**Comment by Rmano** [comment]
More comments: where is the `solar` block defined? I can't find it. 

Moreover, several commits are composed by the shape + fixes to shapes before --- that is confusing. The ideal thing would be to have a single commit for the blocks and a new section in the manual with the blocks here (I do not know how to call them --- TX/RX chain blocks?).

The bias-T block should use the existing shapes of the capacitor and inductor, don't you think? To respect the shape the user prefers for them. 

Finally, the camera, transceiver and power supply blocks are on the edge of "really not a functional block", more a drawing thing. I'll merge them but notice that `circuitikz` scope is to describe circuits (or blocks for circuits), not generic implementation drawings...  so while I think these can be useful, like the oscilloscope shape, I suggest not going more on that way ;-)

Are you ok, I hope, if I reorganize the commits.

**Comment by myzinsky** [comment]
> More comments: where is the solar block defined? I can't find it.

some examples:

https://de.m.wikipedia.org/wiki/Datei:Symbol_Photoelement.svg
https://www.elektronik-kompendium.de/sites/bau/0201112.htm

> Moreover, several commits are composed by the shape + fixes to shapes before --- that is confusing. The ideal thing would be to have a single commit for the blocks and a new section in the manual with the blocks here (I do not know how to call them --- TX/RX chain blocks?).

I'm sorry for that, could you maybe just squash it to one single commit? 

> The bias-T block should use the existing shapes of the capacitor and inductor, don't you think? To respect the shape the user prefers for them.

I was disabled in integrating them, therefore I hope that you could fix that for me :-)

> Finally, the camera, transceiver and power supply blocks are on the edge of "really not a functional block", more a drawing thing. I'll merge them but notice that circuitikz scope is to describe circuits (or blocks for circuits), not generic implementation drawings... so while I think these can be useful, like the oscilloscope shape, I suggest not going more on that way ;-)

Its up to you, if you don't want them we keep them in our local repro :-) 

> Are you ok, I hope, if I reorganize the commits.

Sure please go ahead :-)

**Comment by Rmano** [comment]
> > I think the cross on the SWR meter is not centered --- is it on purpose?
> 
> No I think that's my missing skills on the pgf-base-layer
> 
> > the solar cell should use the opto-arrows like photodiodes, no?
> 
> yes!
> 

Ok, I'll fix them.

> > More comments: where is the solar block defined? I can't find it.
> 
> some examples:
> 
> https://de.m.wikipedia.org/wiki/Datei:Symbol_Photoelement.svg https://www.elektronik-kompendium.de/sites/bau/0201112.htm
> 

No, I mean: ;-)  I can't find the commit that adds that shape...  is it together with some other commit?

> 
> I'm sorry for that, could you maybe just squash it to one single commit?
> 

Don't worry, I'll do that. I'll use it to see if I learn a bit more git... ;-) 

> 
> > Are you ok, I hope, if I reorganize the commits.
> 
> Sure please go ahead :-)

Will do! Thanks!

**Comment by Rmano** [comment]
@myzinsky 

> > More comments: where is the solar block defined? I can't find it.
> 
> some examples:
> 
> 

No, I mean: ;-)  I can't find the commit that adds that shape...  is it together with some other commit?

**Comment by Rmano** [comment]
I will try to merge this branch in my clone, reorganize and then merge from there. I hope I'll not destroy things ;-)

**Comment by Rmano** [comment]
Ok, Now I have it on my own fork. I'll close this when pushing the new one.

**Comment by Rmano** [comment]
Closing, superseded by #838

**Pull request closed without merging**